airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/cursor.py

```diff
@@ -191,18 +190,26 @@ def close_partition(self, partition: Partition) -> None:
         self._has_closed_at_least_one_slice = True
 
     def _add_slice_to_state(self, partition: Partition) -> None:
+        most_recent_record = self._most_recent_record_per_partition.get(partition.stream_name())
+        most_recent_cursor_value = self._extract_cursor_value(most_recent_record) if most_recent_record else None
+
         if self._slice_boundary_fields:
             if "slices" not in self.state:
                 raise RuntimeError(
                     f"The state for stream {self._stream_name} should have at least one slice to delineate the sync start time, but no slices are present. This is unexpected. Please contact Support."
                 )
             self.state["slices"].append(
                 {
-                    "start": self._extract_from_slice(partition, self._slice_boundary_fields[self._START_BOUNDARY]),
-                    "end": self._extract_from_slice(partition, self._slice_boundary_fields[self._END_BOUNDARY]),
+                    self._connector_state_converter.START_KEY: self._extract_from_slice(
+                        partition, self._slice_boundary_fields[self._START_BOUNDARY]
+                    ),
+                    self._connector_state_converter.END_KEY: self._extract_from_slice(
+                        partition, self._slice_boundary_fields[self._END_BOUNDARY]
+                    ),
+                    "most_recent_cursor_value": most_recent_cursor_value,
```

Comment:
Based on the code, this seems odd as the cursor value might be for another partition. It wouldn't necessarily be a problem functionally though, just a bit confusing to see something like:
```
{
  "start": 0,
  "end": 10,
  "most_recent_cursor_value": 1000
}
```

Reply:
Can you describe in what case this scenario is possible? 

To me, it seems quite straightforward. We're adding the slice boundaries and the cursor value from the record that was requested with those boundaries.

Reply:
You are correct; we need to use the partition as a key when saving per partition in this situation. I have also updated the method and the merge_intervals method accordingly